File: minion/app/src/main/java/snakesoft/minion/Activities/SyncActivity.java
package snakesoft.minion.Activities;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import snakesoft.minion.Models.GlobalModel;
import snakesoft.minion.Models.SyncObserver;
import snakesoft.minion.R;

public class SyncActivity extends AppCompatActivity implements SyncListener
{
    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sync);

        final EditText txtView = (EditText) findViewById(R.id.syncText);
        txtView.setEnabled(false);
        GlobalModel.sync(this);
    }

    public void onSyncComplete()
    {
        SyncObserver obs = GlobalModel.getSyncObserver();

        final ProgressBar pb = (ProgressBar) findViewById(R.id.pbSync);
        pb.setVisibility(View.INVISIBLE);

        final EditText txtView = (EditText) findViewById(R.id.syncText);
        txtView.setText(obs.getLog());
    }

    @Override
    public void onSyncError()
    {
        AlertDialog.Builder dlgAlert  = new AlertDialog.Builder(this);
        dlgAlert.setMessage("Sync error: materia unreachable");
        dlgAlert.setTitle("Warning");
        dlgAlert.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                finish();
            }});

        dlgAlert.create().show();
    }
}

File: minion/app/src/main/java/snakesoft/minion/Activities/SyncActivity.kt
package snakesoft.minion.Activities

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import org.jetbrains.anko.*
import org.jetbrains.anko.support.v4._DrawerLayout
import snakesoft.minion.Models.Event1
import snakesoft.minion.Models.GlobalModel
import snakesoft.minion.Models.SyncObserver

class SyncActivity : AppCompatActivity()
{
    override fun onCreate(savedInstanceState: Bundle?)
    {
        val syncObserver = snakesoft.minion.Models.SyncObserver()
        super.onCreate(savedInstanceState)

        val password = intent.getStringExtra("password")

        val onSyncUpdated = Event1<String>()
        SyncActivityUI(onSyncUpdated).setContentView(this)

        doAsync()
        {
            syncObserver.OnUpdated += {
                s -> uiThread() {
                    onSyncUpdated(s)
                }
            }

            GlobalModel.doSync(syncObserver, password)
        }
    }
}

class SyncActivityUI(private val onSyncUpdated: Event1<String>) : MateriaActivityUI<SyncActivity>()
{
    override fun fillActivityUI(_DrawerLayout: @AnkoViewDslMarker _DrawerLayout, ctx: Context)
    {
        with(_DrawerLayout)
        {
            textView()
            {
                singleLine = false
                onSyncUpdated += {
                    s -> append(s)
                    append("\n")
                }
            }
        }
    }
}